Add font-awesome-6 and make it configurable to switch to 5

ursjoss opened the pull request:
Completely migrating to FontAwesome-6 for `wicket-9.x-bootstrap-5.x`.

I initially wanted to fully bump all the way to 6.1.1, but that version is not on maven-central for some reason. So I downgraded to 6.1.0 there. The CSS resource references still point to 6.1.1 though. Not sure what's best to do here... nor why 6.1.1 is not on maven central.

Also pulls along the sample pages, adjusts the footer to not mention glyph-icons but rather font-awesome.

Also downgrades jetty-maven-plugin, as I couldn't run the sample pages without that. Didn't investigate the root cause though.

ursjoss committed:
Adjust versions in CssReferences and pom

ursjoss committed:
Add new icons

ursjoss committed:
update fontawesome-all.css in bootstrap-samples

ursjoss committed:
format

ursjoss committed:
rename classes and adjust icons

ursjoss committed:
Downgrade font-awesome in pom to 6.1.0, as 6.1.1 is not available in maven central

ursjoss committed:
Amend sample page

ursjoss committed:
Migrate footer mention for icons from Glyphicons to Fontawesome

ursjoss committed:
Downgrade jetty-version to run sample page

martin-g commented:
I am not sure whether the removal of 5.x would break someone's application in a minor release.
@solomax Your opinion ?

reckart commented:
The upgrade from FontAwesome 4 to 5 was quite a bit of work in my app. I'm using the abstractions provided by wicket-bootstrap only in some places and in other (most) places, the glyphs are hardcoded into the HTML templates. Being forced to upgrade might be a bit annoying... that said... eventually one will have to upgrade...

solomax commented:
`4 -> 5` upgrade was flawless to me
AFAIR we had an agreement to leave only one version
In such case `5` suffix will only complicate things

@martin-g I would drop suffix :)

reckart commented:
The FA glyph prefixes had changed from 4 to 5. Also some glyphs went away and had to be replaced with others. Finally, widgets such as the bootstrap file upload widget had to be tweaked to work with the new version. :(

solomax commented:
yep :(
tweaks are required :( and it's painful from developer point of view :(

ursjoss commented:
Replacing the FontAwesome 5 classes was what I took out of the discussion in PR #944 - of course only in the `wicket-9.x-bootstrap-5.x` branch, which is not finalized yet. I suggest to leave the Index 6 in the class names nevertheless. That may be helpful in the future, when keeping version 6 and in addition optionally provide version 7 or later of FontAwsome...

But let me know if and how to adjust the PR and I'll have a look.

martin-g commented:
Maybe we should leave both set of classes for each version.
Also the Maven dependencies should become `optional`.
This way the users should decide which version to use in their apps.
Would that work ?

solomax commented:
The code will be duplicated a lot
I thought it would be good idea to have latest versions of everything :)
Sounds like It's better to have icon pack as separate project :)
old ones (4 and 5 for now) will be rarely updated

vrozkovec commented:
If possible, I'd like to be able to pull in FA version of my choosing. For example I'll stick with FA 5 for some time now, because of PRO version that I've purchased with a one-time payment during their Kickstarter campaign. To have FA 6 PRO version you'd need to pay yearly subscription which I kind of don't want to.

ursjoss committed:
Restore FontAwesome5 classes

ursjoss committed:
Make font-awesome an optional dependency

ursjoss commented:
I restored the FontAwesome5 classes and declared the maven depenency optional. Not sure if that captures the stakes of all involved in the discussion.

solomax commented on bootstrap-extensions/pom.xml:
```diff
@@ -21,7 +21,7 @@
         <animate.css.version>3.7.2</animate.css.version>
         <summernote.version>0.8.20</summernote.version>
-        <font-awesome.version>5.15.4</font-awesome.version>
+        <font-awesome.version>6.1.0</font-awesome.version>
```
we should ping someone to release `6.1.1` :)

ursjoss replied:
yes, of course. ~https://github.com/FortAwesome/Font-Awesome/issues/19092~ https://github.com/webjars/font-awesome/issues/70. Thanks, @solomax

ursjoss replied:
font-awesome-6.1.1 is available in maven central now.

solomax commented on pom.xml:
```diff
@@ -62,7 +62,7 @@
         <google-closure.version>v20200830</google-closure.version>
         <javax.javaee-web-api.version>7.0</javax.javaee-web-api.version>
-        <jetty.version>11.0.8</jetty.version>
+        <jetty.version>9.4.44.v20210927</jetty.version>
```
`9.4.46.v20220331`

ursjoss committed:
Bump maven-jetty-plugin to the latest 9.4.X (9.4.46.v20220331)

Thanks @solomax

ursjoss commented on bootstrap-extensions/src/main/java/de/agilecoders/wicket/extensions/markup/html/bootstrap/form/datetime/AbstractDateTimePickerWithIcon.java:
```diff
@@ -130,7 +130,7 @@ protected Component newIcon(String wicketId) {
      */
     protected IconType newIconType() {
-        return FontAwesome5IconType.calendar_alt_r;
+        return FontAwesome6IconType.calendar_r;
```
Here, we'd have a problem if the user uses derivses from this class (or uses (or with `DatetimePickerIconConfig` below) but chose to consume the other variant via maven coordinates, right? (Regardless if we use the 5 or 6 variant, it could go wrong with two versions)

martin-g replied:
I didn't understand the problem here.

ursjoss replied:
Components like this one used to depend on GlyphIcons that were bundled with bootstrap and therefore didn't need an additional dependency. PR #746 in 2018 switched from GlyphIcons to font-awesome icons (at the time to font-awesome-4.7.0, as the only version of font-awesome within wicket-bootstrap. The dependency on org.webjars:font-awesome:4.7.0 was non-optional, so the switch from bundled to non-bundled icons for this and similar components worked flawelessly.

Your initial suggestion for this PR was to make the dependency optional. If a user of wicket-bootstrap would not explicitly depend on font-awesome-6, trying to use the date time picker implementations would not work, as either non font-awesome version would be available at all or only version 5, if that was the users's choice.

The current approach of this PR treats the font-awesome dependency as non-optional, defaulting to version 6. If the user chooses to override the version to keep using version 5, the component selects the appropriate version of the icon.

martin-g commented on bootstrap-extensions/src/main/java/de/agilecoders/wicket/extensions/markup/html/bootstrap/editor/SummernoteEditor.java:
```diff
@@ -90,7 +90,7 @@ public void onComponentTagBody(MarkupStream markupStream, ComponentTag openTag)
     public void renderHead(IHeaderResponse response) {
         response.render(CssHeaderItem.forReference(SummernoteEditorCssReference.instance()));
-        response.render(CssHeaderItem.forReference(FontAwesome5CssReference.instance()));
+        response.render(CssHeaderItem.forReference(FontAwesome6CssReference.instance()));
```
I think we need `IBootstrapSettings#fontAwesomeCssReference` + getter/setter

ursjoss replied:
I did so according to my understanding.

martin-g commented on bootstrap-extensions/src/main/java/de/agilecoders/wicket/extensions/markup/html/bootstrap/form/datetime/DatetimePickerIconConfig.java:
```diff
@@ -15,15 +15,15 @@ public class DatetimePickerIconConfig extends AbstractConfig {
+    private static final IKey<String> Next = newKey("next", FontAwesome6IconType.arrow_right_s.cssClassName());
+    private static final IKey<String> Today = newKey("today", FontAwesome6IconType.calendar_check_r.cssClassName());
+    private static final IKey<String> Clear = newKey("clear", FontAwesome6IconType.eraser_s.cssClassName());
+    private static final IKey<String> Close = newKey("close", FontAwesome6IconType.xmark_s.cssClassName());
```
If we want to support more than one version of FA then we need more complex type in `IBootstrapSettings` than the one I suggested earlier ...
And here those are all `static` fields which makes it harder to even make use of the settings class ...

ursjoss replied:
I wouldn't say it's solved, but at least worked around.

ursjoss commented:
Ok, to summon the plan and align my understanding:

- We extend IBootstrapSettings to accept the FontAwesome{5,6}CdnCssResource (defaulting to 6?)
- CheckboxX, SummernoteEditor, DateTimePicker etc would use the settings to get the reference of choice
- for the problem with the static IKeys in DatetimePickerIconConfig: we could set the default values in the static initialization to something irrelevant and actively set each value in the constructor from the settings...

In any way, IMHO, we'd need to propagate the font-awesome related classe sfrom bootstrap-extensions to bootstrap-core, right?

martin-g commented:
> We extend IBootstrapSettings to accept the FontAwesome{5,6}CdnCssResource (defaulting to 6?)

I think we will need something like `FontAwesomeSettings` that configures the CSS reference, the icon type, ...

martin-g commented:
> In any way, IMHO, we'd need to propagate the font-awesome related classe sfrom bootstrap-extensions to bootstrap-core, right?

Good point! 
IMO we should not move the classes but use the pattern from: 
 https://github.com/apache/wicket/blob/414d3f733e1d07a7f37f88d262b4a828612ab7ec/wicket-native-websocket/wicket-native-websocket-core/src/main/java/org/apache/wicket/protocol/ws/WebSocketSettings.java#L95-L119 (No need of `Holder` class since there are no fields in it).
This way any usage would look like `FontAwesomeSettings.get(Application.get()).getCssReference()`

ursjoss committed:
Revert "Make font-awesome an optional dependency"

This reverts commit 2417c3fc80031af02d49cb50c870fa51bf00aa5e.

ursjoss committed:
Introduce FontAwsomeSettings defaulting to FontAeesome6 but allowing to switch to 5

ursjoss committed:
put concrete version number to README

ursjoss committed:
Sort IconKey values alphabetically

ursjoss committed:
Finalize replacement of Bootstrap6 references with configurable ones

ursjoss commented:
I tried to implement the suggested approach.

As I suggest to use FontAwesome-6 as a default, I removed the optional configuration in maven. We should have a non-optional dependency that can optionally be overriden if the user wants to do that.

I'm not particularly proud nor happy with the map for particular icons used in components inside of `wicket-extensions`. I'm probably overlooking a much simpler solution to allow dynamically picking the icons of the configured version of font-awesome. I'm happy to adjust that if you give me a hint.

ursjoss committed:
Revert "Finalize replacement of Bootstrap6 references with configurable ones"

This reverts commit c79cfe2c26d84e9923d6e09dd3f2b8ef8e2689d4.

ursjoss committed:
Only use BootstrapSettings inside bootstrap-extensions

ursjoss commented:
It is quite a quirk, IMHO.

On second thought, I'd say that `FontAwesomeSettings` should only be used inside of `wicket-extensions` with the exception of overriding the version in the application init method.

Spoiled from using kotlin, I wanted to set FontAwesomeSettings#getIconType and FontAwesomeSettings#IconKey `internal`, only making it available inside the maven module, but obviously that's not that easy in Java.

Client code (such as `wicket-samples`) should probably pick a FontAwesome version (5 or 6) and directly use the respective icons or css references. But wicket-extensions classes that do use font awesome icons can thus reflect the user choice of the version too.

solomax commented on bootstrap-extensions/src/main/java/de/agilecoders/wicket/extensions/markup/html/bootstrap/icon/FontAwesome5CDNCSSReference.java:
```diff
@@ -4,21 +4,21 @@
 import org.apache.wicket.request.resource.UrlResourceReference;
 
 /**
- * <strong>Important</strong>: To use font-awesome 5.x you need to declare the Maven/Gradle dependency
```
these classes are for 5.x
It seems javadoc shouldn't be changed

ursjoss committed:
Add instructions into FontAwesomePage how to switch the version

ursjoss committed:
Fix new solid icons 0-9

ursjoss committed:
refactor getIconName for brand icons

ursjoss committed:
revert some accidential white-space changes

ursjoss committed:
revert some accidential white-space changes

ursjoss commented:
I fixed a bug with solid icons where the enum name starts with `_` (symbols 0-9), added some documentation and tried to remove some inadvertent whitespace changes to at least remove a tiny little bit of noise in the PR.

I will leave the PR as it is until we get some consolidated decision if and how to proceed.

ursjoss committed:
Upgrade font-awesome in pom to 6.1.1, now available in maven central

This reverts commit 263b987b3c661018fb263e32fde28d74a38b7beb.

ursjoss commented:
As we now have font-awsome-6.1.1 in maven central, I think I've addressed all issues. Please comment on how to proceed.

solomax commented:
IMO it's time to merge this PR :)

martin-g commented:
Thanks, @ursjoss !
I will work on it later today!

martin-g committed:
Inline some code to avoid the lookup of the Application thread local

Signed-off-by: Martin Tzvetanov Grigorov <mgrigorov@apache.org>

ursjoss commented:
Thanks to all involved!
